File: Python3/solver.py
# Size of the grid
grid_size = 9

# 0 means unassigned cells
grid = [[2, 0, 7, 0, 1, 0, 4, 0, 6],
        [0, 0, 0, 0, 6, 0, 0, 0, 0],
        [3, 6, 0, 0, 0, 0, 0, 9, 1],
        [4, 0, 0, 6, 0, 5, 0, 0, 7],
        [0, 3, 0, 7, 2, 9, 0, 6, 0],
        [6, 0, 0, 1, 0, 4, 0, 0, 9],
        [1, 4, 0, 0, 0, 0, 0, 7, 3],
        [0, 0, 0, 0, 9, 0, 0, 0, 0],
        [9, 0, 6, 0, 4, 0, 1, 0, 8]]


# Print the grid in the console window
def print_grid(arr):
    for row in range(grid_size):
        for col in range(grid_size):
            print(str(arr[row][col]), end = ' ')
        print()


# Find the next empty cell in the array
def get_next_blank(arr, cell):
    for row_num in range(grid_size):
        for col_num in range(grid_size):
            if arr[row_num][col_num] == 0:
                cell[0] = row_num
                cell[1] = col_num
                return True
    return False


# Check if a cell in the grid is safe to add a specific number
def is_safe(arr, row, col, num):
    # Check if the number already exists in the row
    if num in arr[row]:
        return False

    # Check if the number already exists in the column
    for single_row in arr:
        if single_row[col] == num:
            return False

    # Check if the number already exists in the box
    top_left = [(int(row / 3) * 3), (int(col / 3) * 3)]
    for row_num in range(top_left[0], (top_left[0] + 3)):
        for col_num in range(top_left[1], (top_left[1] + 3)):
            if arr[row_num][col_num] == num:
                return False

    # If it is safe, return true
    return True


# Recursively solve the puzzle using backtracking
def solve_puzzle(arr):
    cell = [0, 0]
    
    # Get the next empty cell, return if there are no empty cells remaining
    if not get_next_blank(arr, cell):
        return True

    # Fill the next empty cell with the next safe digit and recurse
    for num in range(1, 1 + grid_size):
        if is_safe(arr, cell[0], cell[1], num):
            arr[cell[0]][cell[1]] = num

            if solve_puzzle(arr):
                return True
            
            # If we were unable to complete the puzzle using that digit, backtrack the change and empty the cell
            arr[cell[0]][cell[1]] = 0

    # Return false if the puzzle cannot be solved from this point
    return False


# Solve the puzzle and print the result if possible
if solve_puzzle(grid):
    print_grid(grid)
else:
    print("This puzzle cannot be solved.")

File: JavaScript/solver.js
// 0 means unassigned cells
var board =  [[2, 0, 7, 0, 1, 0, 4, 0, 6],
            [0, 0, 0, 0, 6, 0, 0, 0, 0],
            [3, 6, 0, 0, 0, 0, 0, 9, 1],
            [4, 0, 0, 6, 0, 5, 0, 0, 7],
            [0, 3, 0, 7, 2, 9, 0, 6, 0],
            [6, 0, 0, 1, 0, 4, 0, 0, 9],
            [1, 4, 0, 0, 0, 0, 0, 7, 3],
            [0, 0, 0, 0, 9, 0, 0, 0, 0],
            [9, 0, 6, 0, 4, 0, 1, 0, 8]];
            

// Prints the board to the console log
function printBoard(board) {
    let str = "\n";
    for (let rowIndex = 0; rowIndex < 9; rowIndex++) {
        for (let colIndex = 0; colIndex < 9; colIndex++) {
            str = str + board[rowIndex][colIndex] + " ";
        }
        str += "\n";
    }

    console.log(str);
}


// Returns the location of the next empty cell on the board
function findEmptyCell(board) {
    for (let rowIndex = 0; rowIndex < 9; rowIndex++) {
        for (let colIndex = 0; colIndex < 9; colIndex++) {
            if (board[rowIndex][colIndex] == 0) {
                return [rowIndex, colIndex];
            }
        }
    }

    // Return -1 if no empty cells are found
    return [-1, -1];
}


// Checks if the specified value can be entered into the specified cell on the board
function checkValue(board, row, col, value) {
    // Validate row is safe
    if (board[row].includes(value)) {
        return false;
    }

    // Validate column is safe
    for (let rowIndex = 0; rowIndex < board.length; rowIndex++) {
        if (board[rowIndex][col] == value) {
            return false;
        }
    }

    // Validate square is safe
    let squareRow = Math.floor(row / 3) * 3;
    let squareCol = Math.floor(col / 3) * 3;
    for (let rowOffset = 0; rowOffset < 3; rowOffset++) {
        for (let colOffset = 0; colOffset < 3; colOffset++) {
            if (board[squareRow + rowOffset][squareCol + colOffset] == value) {
                return false;
            }
        }
    }

    // Return true if the value passed all tests
    return true;
}


// Solve the board using a backtracking algorithm
function solveBoard(board) {
    let emptyCell = findEmptyCell(board);
    let row = emptyCell[0];
    let col = emptyCell[1];

    // If there are no empty cells, return
    if (row == -1) {
        return board;
    }

    // Try each safe value and recursively solve the board
    for (let value = 1; value <= 9; value++) {
        if (checkValue(board, row, col, value)) {
            board[row][col] = value;
            solveBoard(board);
        }
    }

    // If the board was not solvable from this position, backtrack and clear the cell
    if (findEmptyCell(board)[0] != -1) {
        board[row][col] = 0;
    }

    return board;
}


// Solve the board and print the results
if (solveBoard(board)) {
    printBoard(board);
}
else {
    solveBoard(board);
    console.log("This puzzle cannot be solved.");
}